test/extensions/filters/network/common/fuzz/network_readfilter_fuzz_test.cc

```diff
@@ -9,41 +9,82 @@
 #include "test/fuzz/fuzz_runner.h"
 #include "test/test_common/test_runtime.h"
 
+#include "libprotobuf_mutator/src/libfuzzer/libfuzzer_macro.h"
+#include "src/text_format.h" // from libprotobuf_mutator/
+
 namespace Envoy {
 namespace Extensions {
 namespace NetworkFilters {
 
-DEFINE_PROTO_FUZZER(const test::extensions::filters::network::FilterFuzzTestCase& input) {
+void ensuredValidFilter(unsigned int seed, envoy::config::listener::v3::Filter* config) {
+  // TODO(jianwendong): After extending to cover all the filters, we can use
+  // `Registry::FactoryRegistry<
+  // Server::Configuration::NamedNetworkFilterConfigFactory>::registeredNames()`
+  // to get all the filter names instead of calling `UberFilterFuzzer::filter_names()`.
+  static const auto filter_names = UberFilterFuzzer::filterNames();
+  static const auto factories = Registry::FactoryRegistry<
+      Server::Configuration::NamedNetworkFilterConfigFactory>::factories();
+
+  // Choose a valid filter name.
+  if (std::find(filter_names.begin(), filter_names.end(), config->name()) ==
+      std::end(filter_names)) {
+    absl::string_view filter_name = filter_names[seed % filter_names.size()];
+    if (filter_name != config->name()) {
+      // Clear old config, or unpacking non-suitable value may crash.
+      config->clear_typed_config();
+      config->set_name(std::string(filter_name));
+    }
+  }
+  // Set the corresponding type_url for Any.
+  auto& factory = factories.at(config->name());
+  config->mutable_typed_config()->set_type_url(absl::StrCat(
+      "type.googleapis.com/", factory->createEmptyConfigProto()->GetDescriptor()->full_name()));
+}
+
+static void TestOneProtoInput(const test::extensions::filters::network::FilterFuzzTestCase& input);
+
+using FuzzerProtoType = test::extensions::filters::network::FilterFuzzTestCase;
+// protobuf_mutator::libfuzzer::macro_internal::GetFirstParam<decltype(&TestOneProtoInput)>::type;
+
+extern "C" size_t LLVMFuzzerCustomMutator(uint8_t* data, size_t size, size_t max_size,
+                                          unsigned int seed) {
+  static unsigned config_mutation_cnt = 0;
+  // mutate the config part of the fuzzer only so often.
+  static const unsigned config_mutation_limit = 1000;
+  static protobuf_mutator::Mutator mutator = [seed] {
+    protobuf_mutator::Mutator mutator;
+    mutator.Seed(seed);
+    return mutator;
+  }();
+  static ProtobufMessage::ValidatedInputGenerator generator(
+      seed, ProtobufMessage::composeFiltersAnyMap(), 20);
+  FuzzerProtoType input;
+  input.ParseFromString({reinterpret_cast<const char*>(data), size});
+  mutator.Mutate(input.mutable_actions(), max_size);
+
+  if (config_mutation_cnt > config_mutation_limit) {
```

Comment:
In addition to the high-level comment, I think that either the config or the actions should be modified (not both).

Reply:
Agreed.